File: sort/insert_sort/insert_sort.c
#include <stdio.h>
#include <stdlib.h>
#define ARRAY_SIZE 20

int* insert_sort(int array[], int array_size);
int* create_random_array(void);
void print_array(int* array, int array_size);

/**
 * サイズ20のランダムな配列を作成
 */
int* create_random_array(void) {
  // 長期的に使えるメモリ領域ヒープを確保
  int* array = (int*)malloc(ARRAY_SIZE * sizeof(int));

  srand(3);  // 乱数シード
  for (int i = 0; i < ARRAY_SIZE; i++) {
    // 長い乱数は見にくいので0~99の範囲で乱数を発生させる。
    array[i] = rand() % 100;
  };
  return array;
}

/**
 * insert sort
 */
int* insert_sort(int* array, int array_size) {
  for (int i = 1; i < ARRAY_SIZE; i++) {
    int tmp = array[i];
    int j = i;
    while (j > 0 && array[j - 1] > tmp) {
      array[j] = array[j - 1];
      j--;
    }
    array[j] = tmp;
  }
  return array;
}

/**
 * 1次元配列をprintする
 */
void print_array(int* array, int array_size) {
  // 配列のサイズは配列全体のバイト数 / 要素一つのバイト数で求められる。
  for (int i = 0; i < array_size; i++) {
    printf("%d, ", array[i]);
  }
  printf("\n");
}

int main(void) {
  // ポインタ変数として作成された配列の先頭の値が格納されたメモリの位置を受け取る。
  int* random_array = create_random_array();
  print_array(random_array, ARRAY_SIZE);

  printf("-----insert, sort-----\n");

  int* sorted_array = insert_sort(random_array, ARRAY_SIZE);
  print_array(sorted_array, ARRAY_SIZE);

  // メモリを開放する。
  free(random_array);
  return 0;
}

File: sort/insert_sort/insert_sort.hs
-- 要素を正しい位置に挿入する関数
insert :: (Ord a) => a -> [a] -> [a]
insert e [] = [e] -- 空リストへの挿入
insert e (x : xs)
  | e <= x = e : x : xs -- eをxの左へ
  | otherwise = x : insert e xs -- eをxsの適切な位置へ挿入

-- 挿入ソート関数
insertSort :: (Ord a) => [a] -> [a]
insertSort [] = []
insertSort (x : xs) = insert x (insertSort xs)

main :: IO ()
main = do
  let randomList = [30, 75, 69, 16, 47, 77, 60, 80, 74, 8, 77, 1, 60, 33, 70, 29, 24, 91, 60, 69]
  print randomList
  print $ insertSort randomList
